File: Java/Algorithms/Sorting/Quick_Sort.java
import java.io.*;
import java.util.*;

public class Main {

  public static void quickSort(int[] arr, int lo, int hi) {
    //write your code here
    
    if(lo > hi){
        return;
    }
   int pivot = arr[hi];
   int pi = partition(arr,pivot,lo,hi);
    quickSort(arr,lo,pi -1);
    quickSort(arr,pi+1,hi);
  }

  public static int partition(int[] arr, int pivot, int lo, int hi) {
    System.out.println("pivot -> " + pivot);
    int i = lo, j = lo;
    while (i <= hi) {
      if (arr[i] <= pivot) {
        swap(arr, i, j);
        i++;
        j++;
      } else {
        i++;
      }
    }
    System.out.println("pivot index -> " + (j - 1));
    return (j - 1);
  }

  // used for swapping ith and jth elements of array
  public static void swap(int[] arr, int i, int j) {
    System.out.println("Swapping " + arr[i] + " and " + arr[j]);
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
  }

  public static void print(int[] arr) {
    for (int i = 0; i < arr.length; i++) {
      System.out.print(arr[i] + " ");
    }
    System.out.println();
  }

  public static void main(String[] args) throws Exception {
    Scanner scn = new Scanner(System.in);
    int n = scn.nextInt();
    int[] arr = new int[n];
    for (int i = 0; i < n; i++) {
      arr[i] = scn.nextInt();
    }
    quickSort(arr, 0, arr.length - 1);
    print(arr);
  }

}

File: Rust/Algorithms/src/sort/quick_sort.rs
use super::swap;

fn partition<T: Ord + Copy>(arr: &mut Vec<T>, start: usize, end: usize) -> usize{
    let pivot = arr[end];
    let mut index = start;
    for i in start..end {
        if arr[i] <= pivot {
            swap(arr, i, index);
            index += 1;
        }
    }
    swap(arr, end, index);
    index
}

fn sort<T: Ord + Copy>(arr: &mut Vec<T>, start: usize, end: usize) {
    if start < end {
        let index = partition(arr, start, end);
        sort(arr, start, index - 1);
        sort(arr, index + 1, end);
    }
}

pub fn quick_sort<T: Ord + Copy>(arr: &mut Vec<T>) {
    sort(arr, 0, arr.len() - 1);
}